### ISSUE TITLE
S3 Presign 

### ISSUE THREAD

**Issue by kochie** [m-radzikowski/aws-sdk-client-mock#62]
S3 presign is done using a helper library `@aws-sdk/s3-request-presigner` this seems like something that would need a helper library or additional documentation to mock properly.

It would be good to see a way to mock presign requests. I'm happy to help with this.

**Comment by m-radzikowski** [comment]
I looked at it and I think mocking it is not (easily) possible right now with the lib.

Presign is done by a middleware. `getSignedUrl()` is creating the middleware and makes the `send()` call. Middleware intercepts the command, generates signed URL, and returns the result instead of proceeding with actually sending the command to AWS. See: https://github.com/aws/aws-sdk-js-v3/blob/main/packages/s3-request-presigner/src/getSignedUrl.ts

The mocking lib does not mock middlewares, but only client `send()` command. Mocking middleware can be especially hard in this case, since it's created locally in the `getSignedUrl()` function.

I think the way to go is to mock the whole `getSignedUrl()` function.

Feel free to submit a PR for this!

**Comment by kochie** [comment]
Yeah so at the moment my strategy has been to mock `getSignedUrl()` and this works but it's a pretty basic mock. I'll look into it but might take some time. I'll post here if I have any updates.

**Comment by arunachalamN** [comment]
Yeah, this functionality would be great.

**Comment by ShogunPanda** [comment]
An alternative solution is to mock your URL command to return a `{presigned: require('url').parse('http://url')}`. At least, it worked for me on PutObject.

**Comment by jhecking** [comment]
The best solution I've found is from the [referenced issue](https://github.com/aws/aws-sdk-js-v3/issues/3689#issuecomment-1474822763):

```typescript
const presigner = require('@aws-sdk/s3-request-presigner/dist-cjs/getSignedUrl') // eslint-disable-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment

sinon.stub(presigner, 'getSignedUrl').resolves('https://s3.amazonaws.com/path/to/object')
```

**Comment by pagameba** [comment]
I get an error with this, 

    TypeError: Descriptor for property getSignedUrl is non-configurable and non-writable

**Comment by Roger13** [comment]
This is how I've done using jest:
```
const presigner = require('@aws-sdk/s3-request-presigner/dist-cjs/getSignedUrl')
jest.spyOn(presigner, 'getSignedUrl').mockResolvedValueOnce(signedUrlMockValue)
```
Thanks for the simple solution @jhecking 🙌

**Comment by major-mayer** [comment]
> This is how I've done using jest:
> 
> ```
> const presigner = require('@aws-sdk/s3-request-presigner/dist-cjs/getSignedUrl')
> jest.spyOn(presigner, 'getSignedUrl').mockResolvedValueOnce(signedUrlMockValue)
> ```
> 
> Thanks for the simple solution @jhecking 🙌

This works great, but you need to expose (fake) AWS credentials, or otherwise it will complain.
>     CredentialsProviderError: Could not load credentials from any providers

**Comment by major-mayer** [comment]
Similar behaviour can be achieved for `createPresignedPost` from [s3-presigned-post](https://www.npmjs.com/package/@aws-sdk/s3-presigned-post).
```js
const postPresigner = require('@aws-sdk/s3-presigned-post/dist-cjs/createPresignedPost.js')
jest.spyOn(postPresigner, 'createPresignedPost').mockResolvedValue({ url: 'mockValue',
	fields: {
		key: 'mockValue',
		Policy: 'mockValue',
		'X-Amz-Signature': 'mockValue',
		'X-Amz-Security-Token': 'mockValue',
		bucket: 'mockValue',
		'X-Amz-Algorithm': 'mockValue',
		'X-Amz-Credential': 'mockValue',
		'X-Amz-Date': 'mockValue',
	} })
```

**Comment by Ribosom** [comment]
> This is how I've done using jest:
> 
> ```
> const presigner = require('@aws-sdk/s3-request-presigner/dist-cjs/getSignedUrl')
> jest.spyOn(presigner, 'getSignedUrl').mockResolvedValueOnce(signedUrlMockValue)
> ```
> 
> Thanks for the simple solution @jhecking 🙌

Thank you. I had to change it, but your approach help me.

My solution:

```ts
jest.mock("@aws-sdk/s3-request-presigner", () => {
  return {
    getSignedUrl: jest.fn().mockResolvedValueOnce(signedUrlMockValue),
  };
});
```

**Comment by garysassano** [comment]
> This is how I've done using jest:
> 
> ```
> const presigner = require('@aws-sdk/s3-request-presigner/dist-cjs/getSignedUrl')
> jest.spyOn(presigner, 'getSignedUrl').mockResolvedValueOnce(signedUrlMockValue)
> ```
> 
> Thanks for the simple solution @jhecking 🙌

This solution [doesn't work anymore](https://github.com/aws/aws-sdk-js-v3/issues/5817) starting from SDK v3.495.0 unfortunately.

**Comment by Krrish92** [comment]
Is there any PR open to support getSignedUrl mock ? I am completely blocked as work around discussed above as well not working with new version of sdk

**Comment by m-radzikowski** [comment]
If someone wants to create a PR, I will gladly review it.

**Comment by voslartomas** [comment]
For those still struggling with this issue even today, you can workaround it with reimporting getSignedUrl to your own module  like described in here https://stackoverflow.com/questions/72616800/trying-to-stub-a-function-results-in-descriptor-for-property-is-non-configurable/72793194#72793194

**Comment by oceanofmaya** [comment]
@voslartomas Thanks for pointing to the workaround. It worked for with jest. Since most of my projects uses classes and static methods, I did it this way. I had code that needed to also mock createPresignedPost to do some asserts and it presented the same challenge, and hence I took the same approach for it.

I created  `s3-request-presigner-wrapper-util.ts`

```
import { createPresignedPost as createPresignedPostOriginal } from '@aws-sdk/s3-presigned-post';
import { getSignedUrl as getSignedUrlOriginal } from '@aws-sdk/s3-request-presigner';

/**
 * This is a workaround for the issue where the getSignedUrl method cannot be mocked after v3.495.0.
 * This also applies to createPresignedPost method.
 *
 * Issue: https://github.com/m-radzikowski/aws-sdk-client-mock/issues/62
 * Workaround: https://stackoverflow.com/questions/72616800/trying-to-stub-a-function-results-in-descriptor-for-property-is-non-configurable/72793194#72793194
 *
 * The above workaround allows us to create a wrapper class that can be mocked in tests.
 */
class S3RequestPresignerWrapperUtil {
  static readonly getSignedUrl = getSignedUrlOriginal;
  static readonly createPresignedPost = createPresignedPostOriginal;
}

export { S3RequestPresignerWrapperUtil };
```

This then allowed me to mock it like this and assert where needed

```
const preSignedUrl = 'a-presigned-url';
const presignedPost = { url: 'preSignedPostUrl', fields: {} };

beforeEach(() => {
      jest.resetAllMocks();
      jest.restoreAllMocks();

      jest.spyOn(S3RequestPresignerWrapperUtil, 'getSignedUrl').mockResolvedValueOnce(preSignedUrl);
      jest.spyOn(S3RequestPresignerWrapperUtil, 'createPresignedPost').mockResolvedValueOnce(presignedPost as PresignedPost);
});
